oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+                val message = OASISQueryDTO.newBuilder()
+                        .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
+                                .setName(simName)
+                                .putAllInputVector(inputVector)
+                        )
+                        .build()
+
+                sim.input.onNext(message)
+                val result = select<SimResult> {
+                    sim.output.onReceive { it -> SimResult.Success(it.name, it.outputVectorMap) }
+                    sim.error.onReceive { it ->
+                        SimResult.Failure(it.name, it.exception)
+                    }
+                    if (sim.timeOut != null) {
+                        onTimeout(sim.timeOut.toMillis()) {
+                            //Timed out
+                            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
+                            ).build()
+                            sim.input.onNext(message)
+                            select {
+                                sim.output.onReceive { it -> SimResult.TimeOut(it.name) }
+                                sim.error.onReceive { it -> SimResult.TimeOutFailure(it.name, it.exception) }
+                            }
+                        }
+                    }
+                }
+                when (result) {
+                    is SimResult.Success -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Succeed: Result [${result.result}]"))
+                        viewData.addResult(Result(result.name, "Success", inputVector.toString(), result.result.toString()))
+                    }
+                    is SimResult.Failure -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Failed: Due to\n${result.exception}"))
+                        viewData.addResult(Result(result.name, "Error", inputVector.toString(), "Failure: \n${result.exception}"))
+                    }
+                    is SimResult.TimeOut -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Canceled to timed out after ${sim.timeOut!!.toMillis()} millisecond"))
+                        viewData.addResult(Result(result.name, "Canceled", inputVector.toString(), "Timedout"))
+                    }
+                    is SimResult.TimeOutFailure -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation try to cancel Timed out after ${sim.timeOut!!.toMillis()} millisecond but Failed: Due to\n${result.exception}"))
+                        viewData.addResult(Result(result.name, "Cancel Failed", inputVector.toString(), "Cancellation failed:\n${result.exception}"))
+                    }
+                }
+                i++
```

Comment:
hmm. can we make this implicit somehow or otherwise give it a better name?

Reply:
renamed